src/utils.jl

```diff
@@ -144,4 +160,72 @@ function reduce_tensor(A::AbstractArray{Elt, N}, hyper_index_groups::Array{Array
         Ar[i] = A[map_reduced_indices_to_full(Tuple(i))...]
     end
     Ar
+end
+
+struct BlockTensor{T,N,M} <: AbstractArray{T, N}
+    size::NTuple{N, Int}
+    index_map::NTuple{N, Int}
+    data::AbstractArray{T,M}
+end
+
+BlockTensor(size::NTuple{N, Int64}) where N = BlockTensor{ComplexF64, N}(size)
+
+"""Overload functions from base to make BlockTensor usable"""
+Base.copy(a::BlockTensor{T, N}) where {T, N} = BlockTensor{T, N}(a.size)
+Base.length(a::BlockTensor) = prod(size(a))
+Base.size(a::BlockTensor) = a.size
+Base.getindex(t::BlockTensor, i::Int64) = t[Tuple(CartesianIndices(t.size)[i])...]
+# Base.getindex(t::BlockTensor, i::CartesianIndex) = t[Tuple(i)...]
+Base.IndexStyle(::Type{<:BlockTensor}) = IndexLinear()
+
+"""
+    Base.getindex(t::BlockTensor{T, N, M}, i...) where {T, N, M}
+
+Overload the getindex function. Returns zero if the indiex on ranks identified
+as hyperindex groups differ and the relevant tensor index otherwise.
+"""
+function Base.getindex(t::BlockTensor{T, N, M}, i...) where {T, N, M}
+    new_index = Vector{Int}(undef, M)
+    for x in 1:N
+        xm = findfirst(y -> y == t.index_map[x], t.index_map)
+        if i[x] != i[xm]
+            return convert(T, 0)
+        end
+        new_index[t.index_map[x]] = i[x]
+    end
+    return t.data[new_index...]
+end
+
+Base.show(io::IO, ::MIME"text/plain", a::BlockTensor) =
+    print(io, "BlockTensor with dims $(a.size) and index map $(a.index_map)")
+
+"""
+    expand_tensor(A::AbstractArray{Elt, N}, hyper_index_groups::Array{Int64, 1})
+
+Function to expand the rank of the given tensor assuming the given hyper edge groups.
+Like a generalisation of Diagonal (but returns the full dense tensor). For example
+if passed a vector and given hyperindex groups [1,2], it will return a matrix
+with where non diagonal elements are zero. Efficiency could be improved by creating
```

Comment:
```suggestion
where non diagonal elements are zero. Efficiency could be improved by creating
```

I guess the last sentence was going to refer to a BlockTensor?

Reply:
Yes exactly. This was before I went ahead and implemented. Can update comment to more accurately reflect what's going on now.